apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/InstanceHostManager.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright 2020 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Cloud.Spanner.Admin.Instance.V1;
+using Google.Cloud.Spanner.Common.V1;
+using Google.Protobuf.WellKnownTypes;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Spanner.Data
+{
+    /// <summary>
+    /// Host manager for Spanner data client to get and cache the endpointUris for an instance.
+    /// </summary>
+    public sealed class InstanceHostManager
+    {
+        private ConcurrentDictionary<string, IList<string>> _instanceEndpointMapping =
+            new ConcurrentDictionary<string, IList<string>>();
+
+        /// <summary>
+        /// The default instance host manager.
+        /// </summary>
+        public static InstanceHostManager Default => new InstanceHostManager();
+
+        /// <summary>
+        /// Gets a host for the given instance.
+        /// If resource based routing is enabled (environment variable "GOOGLE_CLOUD_SPANNER_ENABLE_RESOURCE_BASED_ROUTING"),
+        /// this method will attempt to get the given instance endpointUris and return the first one.
+        /// The returning endpointUri will be cached to be used in future inquiries.
+        /// </summary>
+        /// <param name="projectId">The project ID. Must not be <c>null</c>.</param>
+        /// <param name="instanceId">The instance ID. Must not be <c>null</c>.</param>
+        /// <returns>The host for the given instance if found else null.</returns>
+        public async Task<string> GetHostAsync(string projectId, string instanceId)
+        {
+            if (SpannerConstants.IsResourceBasedRoutingEnabled)
+            {
+                if (!_instanceEndpointMapping.ContainsKey(instanceId))
```

Comment:
I was digging into `ChangeDatabase` it clones the `SpannerConnectionStringBuilder` than we need to copy `InstanceHostManager` as well whiling cloning the object.

Other thing that I notice is that while doing `GetOrAddValue` operation it persist the cache of `Host` even after changing the Database name with different Project, in this case we also have to maintain the unique key for `Host`.
WDYT?

Reply:
> I was digging into `ChangeDatabase` it clones the `SpannerConnectionStringBuilder` than we need to copy `InstanceHostManager` as well while cloning the object.

The method `SpannerConnectionStringBuilder.WithDatabase(string database)` does clone the `SpannerConnectionStringBuilder` and that's fine. But the `SpannerConnectionStringBuilder.DatabaseName` property is publicly setable and being a property it couldn't clone the current instance and return the clone even if we wanted to.
But even so, we **don't want to have** a different instance of `InstanceHostManager` per instance of `SpannerConnectionStringBuilder`, we want to have just one cache. Remeber that in my original proposal I wanted `InstanceHostManager` to be a singleton, and @jskeet rightly pointed out that it was better if we had a `Default` instance instead, but this is mostly for testing purposes, so that we can use a mock or a fake of `InstanceHostManager` when unit testing everything else.
So really, the key to the dictionary **needs to be** a compound key of `projectId` and `instanceId`. You can use `InstanceName` which you can easily obtain from `SpannerConnectionStringBuilder` when calling the method `GetHostAsync` method on `InstanceHostManager`.

The dictionary in `InstanceHostManager` would be declared as:
```
private ConcurrentDictionary<InstanceName, IList<string>> _instanceEndpointMapping =
    new ConcurrentDictionary<InstanceName, IList<string>>();
```

The method in InstanceHostManager cloud be declared as:
```
public async Task<string> GetHostAsync(InstanceName instance)
```

And then in SpannerConnectionStringBuilder, on the Host property, you could do something similar to:

```
InstanceName instance = _instanceName ?? instanceName ?? new InstanceName(_databaseName.ProjectId, databaseName.InstanceId);
return GetOrAddValue(nameof(Host), () => Task.Run(() =>
    InstanceHostManager.GetHostAsync(instance))).ResultWithUnwrappedExceptions() ?? SpannerClient.DefaultEndpoint.Host);
```

Reply:
> Other thing that I notice is that while doing `GetOrAddValue` operation it persist the cache of `Host` even after changing the Database name with different Project, in this case we also have to maintain the unique key for `Host`.
> WDYT?

Yes, this is true. I believe you'll have to track if the host is explicitly set by client code either when constructing the `SpannerConnectionStringBuilder` or by calling the `Host` property setter. You can probably add a `bool _explicitHost` for that. If the host has been explicitly set then you don't need to delete the existing  host when the `projectId` or `instanceId` are changed. But if the host has been set by calling `GetHostAsync` or is the default host, then when either `projectId `or `instanceId` change you need to delete the current host so that is set correctly the next time it is requested.
Note: it might be eaiser to code it the other way around, that is track when the host is automatically set.
Let me know if this si clear.

Reply:
@amanda-tarafa As `SpannerConnectionStringBuilder` is internal Property of connection, I don't think database/instance can be change in client code except `ChangeDatabase`.

Implemented `ExplicitHost` logic in `ChangeDatabase` please take a look.

Reply:
> As `SpannerConnectionStringBuilder` is internal Property of connection, I don't think database/instance can be change in client code except `ChangeDatabase`.

Coying from my last comment:

> But even so, we **don't want to have** a different instance of `InstanceHostManager` per instance of `SpannerConnectionStringBuilder`, we want to have just one cache. Remeber that in my original proposal I wanted `InstanceHostManager` to be a singleton, and @jskeet rightly pointed out that it was better if we had a `Default` instance instead, but this is mostly for testing purposes, so that we can use a mock or a fake of `InstanceHostManager` when unit testing everything else.

I'll take a look at the whole thing when tests are passing.

Reply:
For creating mock and other testing it's better to create on `IInstanceManager` interface that have `GetHost` implementation instead of creating `sealed` class.